feat: Employee Referral

Anurag810 opened the pull request:
**Documentation PR:** https://github.com/frappe/erpnext_documentation/pull/300

**Employee Referral**

Workflow:
 1. Create Employee Referral. The initial status will be "Pending".
 2. After submit, you will see a custom button for creating a Job Applicant. 
 3. On click, Job Applicant will be created with status "Open" and the status of Employee Referral document will change to "In Process" 
 4. if someone changes the status of the Job Applicant doc to "Hold" or "Replied", The status of the Employee Referral document will remain ''In Process" 
 5. if the status of the Job Applicant is "Accepted" or "rejected", The status of the Employee Referral document will also change to  "Accepted" or "rejected" respectively.
 6.  After Accepting a job applicant. you can create an additional salary for Refferrer for Compensation, by clicking on create additional salary button. 

Feature:
1. custom button for creating a Job Applicant will be only shown if the status is "Pending".
2.  The status of Employee Referral will always be synchronized with the Job Applicant.
3.  custom button for creating an Additional salary will be only shown if the status is "Accepted" and the checkbox "Is applicable for employee referral " is checked.
4. You can create only one additional salary for one Employee Referral document.  
 

Form and dashboard: 
![image](https://user-images.githubusercontent.com/32095923/112471530-07c49f00-8d92-11eb-8145-4c1d52168e00.png)

List view indicators:
![image](https://user-images.githubusercontent.com/32095923/112471816-5eca7400-8d92-11eb-8056-b26499ef81b1.png)

Additional Salary:
![image](https://user-images.githubusercontent.com/32095923/112471972-90dbd600-8d92-11eb-90eb-53c27ceaf3f9.png)

Anurag810 committed:
feat: Employee referal design and status sync

Anurag810 committed:
feat: create job Applicant and linked Document

Anurag810 committed:
feat: Added list view indicator

Anurag810 committed:
chore: formatted file

Anurag810 committed:
feat: Addedd Employee Referral form Dashboard

Anurag810 committed:
feat: pay compensation via additional salary

Anurag810 committed:
test: Employee Rreferral

ruchamahabal commented:
- Fix Translation Syntax
- Fix Sider Issues
- Add documentation
- Check Failing Test
   
   ![image](https://user-images.githubusercontent.com/24353136/114691790-a90d9800-9d35-11eb-8540-43b5665f36b5.png)

ruchamahabal commented on erpnext/hr/doctype/employee_advance/employee_advance.js:
```diff
@@ -2,19 +2,19 @@
 		frm.add_fetch("company", "default_employee_advance_account", "advance_account");
 
-		frm.set_query("employee", function() {
+		frm.set_query("employee", function () {
```
```suggestion
		frm.set_query("employee", function() {
```

We don't need a space between the call and the function name. Update wherever changed

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.js:
```diff
@@ -0,0 +1,52 @@
+
+
+
+	},
+	create_job_applicant: function(frm) {
```
```suggestion
	},

	create_job_applicant: function(frm) {
```

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.js:
```diff
@@ -0,0 +1,52 @@
+
+frappe.ui.form.on('Employee Referral', {
+	refresh: function(frm) {
+		if (frm.doc.docstatus == 1 && frm.doc.status === "Pending") {
```
```suggestion
		if (frm.doc.docstatus === 1 && frm.doc.status === "Pending") {
```

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.js:
```diff
@@ -0,0 +1,52 @@
+
+			let additional_salary_count = data[0].count;
+
+			if (frm.doc.docstatus == 1 &&frm.doc.status === "Accepted" &&
```
Move this check for status before the additional salary check since we should only check for additional salary if the status is **Accepted**.

Also, it throws this error on form load:

![image](https://user-images.githubusercontent.com/24353136/114733512-079c3b80-9d61-11eb-949f-3d89127c8a5e.png)

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.json:
```diff
@@ -0,0 +1,274 @@
+   "no_copy": 1,
+   "options": "Pending\nIn Process\nAccepted\nRejected",
+   "permlevel": 1,
+   "read_only": 1,
```
```suggestion
```
Let's not keep this status as read only. Some people might not create a Job Applicant at all and would want to Accept or Reject from here only.

Anurag810 replied:
Made it read-only because the status of **Employee referral** should be(always) tracked by the **job applicant** status. if a user wants to cancel a referral, have to cancel a document... no need to create a job applicant in the first place. otherwise, status is synced accordingly. the two-way status syncing is an overhead.

ruchamahabal replied:
But if someone doesn't want to create a Job Applicant, they would want to at least Reject Referral from here itself. In that case, we should check if a job applicant exists against the doc, if not, then show a custom button to change the status.

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.py:
```diff
@@ -0,0 +1,59 @@
+		self.set_full_name()
+
+	def set_full_name(self):
+		self.full_name = self.first_name + " "+self.last_name
```
```suggestion
		self.full_name = self.first_name + " " + self.last_name
```

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.py:
```diff
@@ -0,0 +1,59 @@
+
+	#just for Api call if some set status apart from default Status
+	status = emp_ref.status
+	if emp_ref.status in ["Pending", "In process"]:
```
```suggestion
	if emp_ref.status in ["Pending", "In Process"]:
```

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.py:
```diff
@@ -0,0 +1,59 @@
+
+	job_applicant.save()
+	frappe.msgprint(_("Job Applicant created successfully. {0}").format(get_link_to_form("Job Applicant", job_applicant.name)))
+	emp_ref.db_set('status', "In Process")
```
```suggestion
	emp_ref.db_set("status", "In Process")
```

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.py:
```diff
@@ -0,0 +1,59 @@
+	if isinstance(doc, string_types):
+		doc = frappe._dict(json.loads(doc))
+
+	if not frappe.db.exists("Additional Salary", {'ref_docname': doc.name}):
```
```suggestion
	if not frappe.db.exists("Additional Salary", {"ref_docname": doc.name}):
```

Also, add a server-side message for the else part to communicate that Additional Salary already exists.

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.py:
```diff
@@ -0,0 +1,59 @@
+		doc = frappe._dict(json.loads(doc))
+
+	if not frappe.db.exists("Additional Salary", {'ref_docname': doc.name}):
+		additional_salary = frappe.new_doc('Additional Salary')
```
```suggestion
		additional_salary = frappe.new_doc("Additional Salary")
```

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral_dashboard.py:
```diff
@@ -0,0 +1,16 @@
+		},
+		'transactions': [
+			{
+				'items': ["Job Applicant", "Additional Salary"]
```
```suggestion
				'items': ['Job Applicant', 'Additional Salary']
```

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.json:
```diff
@@ -0,0 +1,274 @@
+   "label": "Referrer Details"
+  },
+  {
+   "fieldname": "employee",
+   "fieldtype": "Link",
+   "label": "Employee ",
+   "options": "Employee",
+   "reqd": 1
```
```suggestion
   "fieldname": "referred_by",
   "fieldtype": "Link",
   "label": "Referred By ",
   "options": "Employee",
   "reqd": 1
```
I think we should label this field as Referred By instead of Employee.

Anurag810 replied:
As there is already a separate section for Referral(Refer to screenshot). and if we name it Referred By... what should be a label for **Employee name**...
the employee is pretty standard. but maybe we can rename it to Referrer and Referrer name..
![image](https://user-images.githubusercontent.com/32095923/116040912-392ed400-a68a-11eb-9fdd-0646bd9b5faf.png)

ruchamahabal replied:
Yes, let's rename it to Referrer and Referrer Name

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.json:
```diff
@@ -0,0 +1,274 @@
+  },
+  {
+   "default": "1",
+   "fieldname": "is_applicable_for_referral_compensation",
+   "fieldtype": "Check",
+   "label": "Is applicable for referral compensation"
```
```suggestion
   "fieldname": "is_applicable_for_referral_bonus",
   "fieldtype": "Check",
   "label": "Is Applicable for Referral Bonus"
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.json:
```diff
@@ -0,0 +1,274 @@
+   "label": "Additional Information "
+  },
+  {
+   "fieldname": "why_this_candidate_is_qualified_for_this_position",
+   "fieldtype": "Text",
+   "label": "Why this Candidate is qualified for this position"
```
```suggestion
   "fieldname": "qualification_reason",
   "fieldtype": "Text",
   "label": "Why is this Candidate Qualified for this Position?"
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.json:
```diff
@@ -0,0 +1,274 @@
+   "read_only": 1
+  },
+  {
+   "collapsible": 1,
```
```suggestion
```
Let's not keep this collapsible. The referrer might not fill this up if it is collapsed but this data could be useful while capturing referrals

Anurag810 committed:
Update erpnext/hr/doctype/employee_advance/employee_advance.js

Co-authored-by: Rucha Mahabal <ruchamahabal2@gmail.com>

Anurag810 committed:
fix: changes requested

Anurag810 committed:
fix: changes requested

Anurag810 committed:
fix: sider

Anurag810 committed:
fix: translation

Anurag810 committed:
fix: test

Anurag810 committed:
feat: added to Dashboard

Anurag810 commented:
@ruchamahabal test passed.
![image](https://user-images.githubusercontent.com/32095923/116057441-6d12f500-a69c-11eb-9313-a6a994304520.png)

ruchamahabal commented on erpnext/hr/doctype/employee_referral/employee_referral.json:
```diff
@@ -0,0 +1,273 @@
+   "default": "1",
+   "fieldname": "is_applicable_for_referral_bonus",
+   "fieldtype": "Check",
+   "label": "Is applicable for referral Bonus"
```
```suggestion
   "label": "Is Applicable for Referral Bonus"
```

Anurag810 committed:
fix: changes

Anurag810 commented:
@ruchamahabal 
docs: https://github.com/frappe/erpnext_documentation/pull/300

ruchamahabal committed:
chore: clean-up Employee Referral form

- fix labels

- fix full name field

- fix failing test

- set title for Employee Referral form

- set fields in List View and Standard Filter

ruchamahabal committed:
feat: option to add a resume link

- copy resume and resume link from Employee Referral to Job Applicant

ruchamahabal committed:
fix: multiple fixes

- confirm before rejecting employee referral

- set primary actions for custom buttons

- Validations for additional salary against employee referrals with status as accepted only

- fix list view indicators

- code formatting style

ruchamahabal committed:
feat: Add field to track Referral Bonus Payment Status

- fix visibility of additional salary button

ruchamahabal committed:
Merge branch 'develop' into Employee-Referrals

ruchamahabal commented:
All tests are passing:

![image](https://user-images.githubusercontent.com/24353136/116121010-cd765680-a6dd-11eb-9e31-e9642a017d87.png)

But submitting the coverage is failing which is unrelated to this PR, hence merging.
